### PR TITLE
Listen and publish database changes upstream

### LINKED ISSUE #74: Attach listener for syncing changes between shared plans and trips

### PR DESCRIPTION
**Pull request by keanecjy**
Utilize delegate to pass data from repository --> service --> viewmodel

This updates the @Published object in the view model, which then updates the view

Closes #74

### CODE REVIEW

File: TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift [hunk 1]
```diff
@@ -98,12 +74,11 @@ class PlanViewModel<T: Plan>: ObservableObject {
...
 
         guard hasUpdatedPlan, planErrorMessage.isEmpty else {
-            self.isLoading = false
-            self.hasError = true
+            handleError()
             return
         }
 
-        await updatePublishedProperties(plan: updatedPlan)
+        // await updatePublishedProperties(plan: updatedPlan)
```
Comment (tanruiquan) [inline TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift]: Is there a reason why this is being commented out?

File: TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift [hunk 2]
```diff
@@ -47,34 +47,11 @@ class PlanViewModel<T: Plan>: ObservableObject {
...
-            self.isLoading = false
-            self.hasError = true
-            return
-        }
-
-        await updatePublishedProperties(plan: plan)
-
-        self.isLoading = false
+        self.isLoading = true
+        await planService.fetchPlanAndListen(withPlanId: plan.id)
```
Comment (tanruiquan) [inline TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift:54]: I understand that `planService` will call the delegate's `update()` function which have a `self.isLoading = false` but it was quite hard to trace and follow the logic. Is is possible to put the change here like this?
```suggestion
        self.isLoading = true
        await planService.fetchPlanAndListen(withPlanId: plan.id)
        self.isLoading = false
```

**Review by tanruiquan — APPROVED** [review]
Just a few comments, LGTM 👍

File: TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift (hunk 1 above)
Comment (keanecjy) [reply TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift]: Forgot to remove this, now `updatePublishedProperties` be called from the `update` callback function after firebase publishes the changes

File: TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift (hunk 2 above)
Comment (keanecjy) [reply TourMate/TourMate/Frontend/Plan/Views/PlanViews/PlanViewModel.swift:54]: Ok, I think this is fine.

**Pull request merged**